packages/fast-usdc/src/exos/liquidity-pool.js

```diff
@@ -89,60 +111,154 @@ export const prepareLiquidityPoolKit = (zone, zcf, USDC, tools) => {
+        outstandingLends,
+        poolStats,
+        poolMetricsRecorderKit,
         poolSeat,
         PoolShares,
         proposalShapes,
-        shareWorthRecorderKit,
+        shareMint,
+        shareWorth,
       };
     },
     {
-      feeSink: {
+      borrower: {
+        lookupBalance() {
+          const { poolSeat } = this.state;
+          return poolSeat.getAmountAllocated('USDC', usdc.brand);
+        },
         /**
-         * @param {Amount<'nat'>} amount
-         * @param {Payment<'nat'>} payment
+         * @param {ZCFSeat} assetManagerSeat
+         * @param {{ USDC: Amount<'nat'>}} amounts
          */
-        async receive(amount, payment) {
-          const { poolSeat, shareWorth } = this.state;
-          const { external } = this.facets;
-          await depositToSeat(
-            zcf,
-            poolSeat,
-            harden({ USDC: amount }),
-            harden({ USDC: payment }),
+        borrow(assetManagerSeat, amounts) {
+          const { outstandingLends, poolSeat, poolStats } = this.state;
+          const available = poolSeat.getAmountAllocated('USDC', usdc.brand);
+          // Validate amount is available in pool
+          const post = borrowCalc(
+            amounts.USDC,
+            available,
+            outstandingLends,
+            poolStats,
           );
-          this.state.shareWorth = withFees(shareWorth, amount);
-          external.publishShareWorth();
+
+          // COMMIT POINT
+          try {
+            zcf.atomicRearrange(
+              harden([[poolSeat, assetManagerSeat, amounts]]),
+            );
+          } catch (cause) {
+            const reason = Error('🚨 cannot commit borrow', { cause });
+            console.error(reason.message, cause);
+            zcf.shutdownWithFailure(reason);
+          }
+
+          Object.assign(this.state, post);
+          this.facets.external.publishPoolMetrics();
         },
+        // TODO repay failed `LOA.deposit()`
       },
+      repayer: {
+        /**
+         * @param {ZCFSeat} assetManagerSeat
+         * @param {RepayAmountKWR} amounts
+         */
+        repay(assetManagerSeat, amounts) {
+          const {
+            contractSeat,
+            outstandingLends,
+            poolSeat,
+            poolStats,
+            shareWorth,
+          } = this.state;
+          checkPoolBalance(poolSeat, shareWorth, usdc.brand, outstandingLends);
+          // determine new shareWorth, outstandingLends, and poolMetrics
+          const post = repayCalc(
+            shareWorth,
+            amounts,
+            outstandingLends,
+            poolStats,
+          );
+
+          const { ContractFee, ...rest } = amounts;
+
+          // COMMIT POINT
+          try {
+            zcf.atomicRearrange(
+              harden([
+                [
+                  assetManagerSeat,
+                  poolSeat,
+                  rest,
+                  { USDC: add(amounts.PoolFee, amounts.Principal) },
```

Comment:
Add a test where the `fromSeat` doesn't have enough to cover these amounts, please? I think the `atomicRearrange` will fail, taking the contract down with it.

It looks like we should push this `add()` down into `repayCalc`. And `repayCalc` should get the seat allocation and the amounts. This isn't a zoe offer handler, so we shouldn't rely on the caller to ensure that the seat has enough.

Reply:
Please see "repay fails when amounts do not match seat allocation" and "repay fails when seat allocation does not equal amounts". 

I did not push the `add()` into `repayCalc`, but believe we can confidence `atomicRearrange` will not fail. There are additional tests that show the AKWR guards working as expected.